### PR TITLE
Darkmeiro decompilation

### PR DESCRIPTION
**Pull request by petrie911**
Bg_Gnd_Darkmeiro decompiled, matched, and documented.

### CODE REVIEW

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 1]
```diff
@@ -35,25 +35,225 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+
+//#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/BgGndDarkmeiro_NoAction.s")
+
+void BgGndDarkmeiro_PlatformTimer(BgGndDarkmeiro *this, GlobalContext *globalCtx) {
+    s16 phi_a1;
+
+    if (Flags_GetSwitch(globalCtx, ((this->dyna.actor.params >> 8) & 0x3F) + 1) != 0) { //If switch N+1 is active...
+        if ((this->actorState & 4) != 0) { // and timer 1 is running...
+            if (this->timer1 > 0) {//if there's any time left...
+                this->timer1 -= 1; //decrement timer 1.
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:144]: This is very minor but in C you can use the `--` operator.
```suggestion
                this->timer1--; //decrement timer 1.
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 2]
```diff
@@ -35,25 +35,225 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+			this->timer2 =  (Math_Coss( (s16) newvar2 * 4096 ) * 64.0f) + 127.0f;
+			if (this->timer2 >= 0x80) {
+				this->timer2 = 0x7F;
+			}
+        } else {
+            this->timer2 = newvar * 8;  
+        } 		//This block of code sets the transparency of the platforms and causes them to flash before they vanish when their switch is unset. Or it would if they didn't ignore all of their color data.
+		
+		gfxCtx = globalCtx->state.gfxCtx;
+        Graph_OpenDisps(&dispRefs, globalCtx->state.gfxCtx,"../z_bg_gnd_darkmeiro.c", 0x17A);
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:240]: We use decimal for line numbers.
```suggestion
        Graph_OpenDisps(&dispRefs, globalCtx->state.gfxCtx,"../z_bg_gnd_darkmeiro.c", 378);
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h [hunk 3]
```diff
@@ -10,7 +10,10 @@ typedef void (*BgGndDarkmeiroActionFunc)(struct BgGndDarkmeiro*, GlobalContext*)
 
 typedef struct BgGndDarkmeiro {
     /* 0x0000 */ DynaPolyActor dyna;
-    /* 0x0164 */ char unk_164[0x8];
+    /* 0x0164 */ u16 actorState; // Bitflags for whether a platform or timer is on. 0x8 for timer 2, 0x4 for timer 1, 0x2 for whether a platform is on.
+	/* 0x0166 */ s16 timer1; //Timer for switch flag N+1. Platforms use it to count down to vanishing after the switch flag is unset.
+	/* 0x0168 */ s16 timer2; //Timer for switch flag N+2. Platforms use it to control their transparency (does not work).
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h:15]: Since the format.sh script won't format .h files, you'll have to make sure you ident the headers right manually.
```suggestion
    /* 0x0164 */ u16 actorState; // Bitflags for whether a platform or timer is on. 0x8 for timer 2, 0x4 for timer 1, 0x2 for whether a platform is on.
    /* 0x0166 */ s16 timer1; //Timer for switch flag N+1. Platforms use it to count down to vanishing after the switch flag is unset.
    /* 0x0168 */ s16 timer2; //Timer for switch flag N+2. Platforms use it to control their transparency (does not work).
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h [hunk 4]
```diff
@@ -10,7 +10,10 @@ typedef void (*BgGndDarkmeiroActionFunc)(struct BgGndDarkmeiro*, GlobalContext*)
...
+	/* 0x016A */ s16 unk16A; //unused
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h:16]: `unk16A` here actually looks like padding.
```suggestion
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 5]
```diff
@@ -35,25 +35,225 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/func_80879134.s")
+	newvar = this->timer1;
+    if (newvar != 0) { // Platform only gets drawn if vanishing timer isn't zero.
+        if (newvar >= 0x41) {
+			this->timer2 = (this->timer2 < 0x78) ? this->timer2 + 8 : 0x7F;
+        } else if (newvar >= 0x11) {
+			newvar2 = this->timer1 ;
+			this->timer2 =  (Math_Coss( (s16) newvar2 * 4096 ) * 64.0f) + 127.0f;
+			if (this->timer2 >= 0x80) {
+				this->timer2 = 0x7F;
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:233]: This doesn't matter for matching but I believe `>` makes more sense than `>=` here according to the values.
```suggestion
        if (newvar > 0x40) {
			this->timer2 = (this->timer2 < 0x78) ? this->timer2 + 8 : 0x7F;
        } else if (newvar > 0x10) {
			newvar2 = this->timer1 ;
			this->timer2 =  (Math_Coss( (s16) newvar2 * 4096 ) * 64.0f) + 127.0f;
			if (this->timer2 > 0x7F) {
				this->timer2 = 0x7F;
```

File: spec [hunk 6]
```diff
@@ -718,7 +718,6 @@ endseg
 beginseg
     name "ovl_Bg_Gnd_Darkmeiro"
     include "build/src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.o"
-    include "build/data/overlays/actors/z_bg_gnd_darkmeiro.data.o"
     include "build/data/overlays/actors/z_bg_gnd_darkmeiro.reloc.o"
```
Comment (Random06457) [inline spec:721]: Once you get an overlay to fully match you can delete the old reloc file (`data/overlays/actors/z_bg_gnd_darkmeiro.reloc.s` in your case) and use the one auto generated.
```suggestion
    include "build/src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/ovl_Bg_Gnd_Darkmeiro_reloc.o"
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 7]
```diff
@@ -35,25 +35,225 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
 };
-*/
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/func_80878C30.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/BgGndDarkmeiro_Init.s")
+extern UNK_TYPE D_0600C080;
+extern UNK_TYPE D_060088B0;
+extern UNK_TYPE D_0600BEC0;
+
+void BgGndDarkmeiro_TogglePlatform(BgGndDarkmeiro *this, GlobalContext *globalCtx) {
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:43]: We use the `type* name` syntax for pointers. Make sure to run `format.sh` before making a PR so your files are formatted accordingly to our style conventions.
```suggestion
void BgGndDarkmeiro_TogglePlatform(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 8]
```diff
@@ -35,25 +34,209 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+}
+
+void BgGndDarkmeiro_DrawStaticPlatform(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
+    GraphicsContext* gfxCtx = globalCtx->state.gfxCtx;
+    Gfx* dispRefs[4];
+
+    Graph_OpenDisps(&dispRefs, globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 391);
+    gDPSetPrimColor(gfxCtx->polyXlu.p++, 0, 0, 0xC6, 0xCA, 0xD0, 0xFF);
+    Graph_CloseDisps(&dispRefs, globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 393);
+    Gfx_DrawDListXlu(globalCtx, &D_0600BEC0); // Draws the platform.
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:241]:
```suggestion
    Graph_OpenDisps(dispRefs, globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 391);
    gDPSetPrimColor(gfxCtx->polyXlu.p++, 0, 0, 0xC6, 0xCA, 0xD0, 0xFF);
    Graph_CloseDisps(dispRefs, globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 393);
    Gfx_DrawDListXlu(globalCtx, D_0600BEC0); // Draws the platform.
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 9]
```diff
@@ -35,25 +34,209 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+        } else {
+            this->timer2 = vanishTimer * 8;
+        } // This block of code sets the transparency of the platforms and causes them to flash before they
+          // vanish when their switch is unset. Or it would if they didn't ignore all of their color data.
+
+        gfxCtx = globalCtx->state.gfxCtx;
+        Graph_OpenDisps(&dispRefs, globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 378);
+        gDPSetPrimColor(gfxCtx->polyXlu.p++, 0, 0, 0xC6, 0xCA, 0xD0, this->timer2);
+        Graph_CloseDisps(&dispRefs, globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 380);
+        Gfx_DrawDListXlu(globalCtx, &D_0600BEC0); // Draws the platform.
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:230]:
```suggestion
        Graph_OpenDisps(dispRefs, globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 378);
        gDPSetPrimColor(gfxCtx->polyXlu.p++, 0, 0, 0xC6, 0xCA, 0xD0, this->timer2);
        Graph_CloseDisps(dispRefs, globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 380);
        Gfx_DrawDListXlu(globalCtx, D_0600BEC0); // Draws the platform.
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 10]
```diff
@@ -35,25 +34,209 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/BgGndDarkmeiro_Update.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/func_808791BC.s")
-
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/func_808791E8.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/func_80879334.s")
+extern UNK_TYPE D_0600C080;
+extern UNK_TYPE D_060088B0;
+extern UNK_TYPE D_0600BEC0;
```
Comment (Random06457) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:40]: `D_0600BEC0` and `D_060088B0` are dlists here.
```suggestion
extern Gfx D_060088B0[];
extern Gfx D_0600BEC0[];
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 11]
```diff
@@ -35,25 +34,196 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/func_808791E8.s")
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/actors/ovl_Bg_Gnd_Darkmeiro/func_80879334.s")
+extern Gfx D_060088B0[];
+extern Gfx D_0600BEC0[];
+
+extern UNK_TYPE D_0600C080;
+
+void BgGndDarkmeiro_ToggleBlock(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
+
```
Comment (ethteck) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:44]: This empty line is unnecessary. Was format.sh run on this?

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 12]
```diff
@@ -35,25 +34,196 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+            break;
+    }
+    if (globalCtx) {}; // needed for matching
+}
+
+void BgGndDarkmeiro_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    BgGndDarkmeiro* this = THIS;
+
+    if ((this->dyna.actor.params & 0xFF) == 1) {
+        if (!globalCtx) {}; // needed for matching
```
Comment (ethteck) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:115]: you probably don't need the ! in front of globalCtx, at least

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 13]
```diff
@@ -35,25 +34,196 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+    }
+}
+
+void BgGndDarkmeiro_DoNothing(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
+}
+
+void BgGndDarkmeiro_BlockTimer(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
+    s16 timeLeft;
+
+    if (Flags_GetSwitch(globalCtx, ((this->dyna.actor.params >> 8) & 0x3F) + 1) != 0) {
```
Comment (ethteck) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:126]: this function is a bool so you don't need the `!= 0`

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 14]
```diff
@@ -35,25 +34,196 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+                this->actionState &= ~4;
+            }
+        } else {
+            this->actionState |= 4;
+            this->timer1 = 304;
+            Audio_PlaySoundGeneral(NA_SE_EV_RED_EYE, &D_801333D4, 4, &D_801333E0, &D_801333E0, &D_801333E8);
+        }
+    }
+
+    if (Flags_GetSwitch(globalCtx, ((this->dyna.actor.params >> 8) & 0x3F) + 2) != 0) {
```
Comment (ethteck) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:141]: here too

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 15]
```diff
@@ -35,25 +34,196 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+    } else {
+        Flags_UnsetSwitch(globalCtx, ((this->dyna.actor.params >> 8) & 0x3F));
+    }
+}
+
+void BgGndDarkmeiro_StaticBlock(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
+}
+
+void BgGndDarkmeiro_SwitchBlock(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
+
```
Comment (ethteck) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:171]: extra empty line

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 16]
```diff
@@ -35,25 +34,196 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+}
+
+void BgGndDarkmeiro_Update(Actor* thisx, GlobalContext* globalCtx) {
+    BgGndDarkmeiro* this = THIS;
+
+    this->actionFunc(this, globalCtx);
+}
+
+void BgGndDarkmeiro_Draw0(Actor* thisx, GlobalContext* globalCtx) {
+
```
Comment (ethteck) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:190]: here 2

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 17]
```diff
@@ -35,25 +34,196 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+    Gfx_DrawDListXlu(globalCtx, D_060088B0);
+}
+
+void BgGndDarkmeiro_DrawSwitchBlock(Actor* thisx, GlobalContext* globalCtx) {
+    BgGndDarkmeiro* this = THIS;
+    s16 vanishTimer;
+    u16 vanishTimerU;
```
Comment (ethteck) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:197]: Have you tried removing this temp? I doubt it's necessary and seems weird for someone to create and use it like it's currently being used.

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 18]
```diff
@@ -35,25 +34,196 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+        gDPSetPrimColor(oGfxCtx->polyXlu.p++, 0, 0, 0xC6, 0xCA, 0xD0, this->timer2);
+        //@ bug: Due to a bug in the display list, the transparency data is not used.
+        CLOSE_DISPS(globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 380);
+
+        Gfx_DrawDListXlu(globalCtx, D_0600BEC0);
+    }
+}
+
+void BgGndDarkmeiro_DrawStaticBlock(Actor* thisx, GlobalContext* globalCtx) {
+
```
Comment (ethteck) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:223]: Empty line

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c (hunk 11 above)
Comment (petrie911) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c]: yes. it doesn't seem to mind them.

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 19]
```diff
@@ -13,17 +13,16 @@
...
 
-void func_80878C30(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_80878F28(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_80878F34(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_80879128(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_80879134(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_808791BC(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_808791E8(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_80879334(BgGndDarkmeiro* this, GlobalContext* globalCtx);
+void BgGndDarkmeiro_ToggleBlock(BgGndDarkmeiro* this, GlobalContext* globalCtx);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:20]: This prototype can be omitted, the function's body is declared before being used

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 20]
```diff
@@ -35,25 +34,189 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+
+void BgGndDarkmeiro_Init(Actor* thisx, GlobalContext* globalCtx) {
+    BgGndDarkmeiro* this = THIS;
+    s32 local_c = 0;
+    s32 pad;
+
+    this->actionFunc = BgGndDarkmeiro_DoNothing;
+    Actor_SetScale(&this->dyna.actor, 0.1f);
+
+    switch (thisx->params & 0xFF) {
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:63]: Can you use `this->dyna.actor.params` instead of through `thisx`?

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 21]
```diff
@@ -35,25 +34,189 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+    }
+}
+
+void BgGndDarkmeiro_DoNothing(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
+}
+
+void BgGndDarkmeiro_BlockTimer(BgGndDarkmeiro* this, GlobalContext* globalCtx) {
+    s16 timeLeft;
+
+    if (Flags_GetSwitch(globalCtx, ((this->dyna.actor.params >> 8) & 0x3F) + 1)) {
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:124]: `(this->dyna.actor.params >> 8) & 0x3F` is the same as `this->dyna.actor.params >> 8 & 0x3F` (you get used to it and it's a lot easier to read)

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 22]
```diff
@@ -35,25 +34,189 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+            this->timer2 = (Math_Coss((u16)this->timer1 * 4096) * 64.0f) + 127.0f;
+            if (this->timer2 > 127) {
+                this->timer2 = 127;
+            }
+        } else {
+            this->timer2 = vanishTimer * 8;
+        }
+
+        OPEN_DISPS(globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 378);
+        gDPSetPrimColor(oGfxCtx->polyXlu.p++, 0, 0, 0xC6, 0xCA, 0xD0, this->timer2);
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:208]: Use decimal for colors

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h [hunk 23]
```diff
@@ -10,10 +10,26 @@ typedef void (*BgGndDarkmeiroActionFunc)(struct BgGndDarkmeiro*, GlobalContext*)
...
+    /* 0x0168 */ s16 timer2;         /* Timer for switch flag N+2. Blocks use it to control
+                                        their transparency (does not work).                 */
     /* 0x016C */ BgGndDarkmeiroActionFunc actionFunc;
 } BgGndDarkmeiro; // size = 0x0170
 
+typedef enum {
+    /* 0 */ MODE_ZERO,   // This does nothing and has a display list which apparently draws nothing.
+    /* 1 */ CLEAR_BLOCK, /* Clear blocks appear when their switch flag is set and disappear 64 frames
+                            after their flag is cleared. Clear blocks with flag 0x3F are always on.   */
+    /* 2 */ BLOCK_TIMER  /* A block timer with switch flag N reacts to switch flags N+1 and N+2 being
```
Comment (Dragorn421) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h:26]: You want to prefix those with eg `DARKMEIRO_TYPE_` so they don't clutter the global symbol list

**Review by Dragorn421 — CHANGES_REQUESTED** [review]
Small things

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c (hunk 21 above)
Comment (petrie911) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:124]: I don't have C operator preference memorized, so I'd prefer to keep the (). Also this should just be SWITCH_FLAG or something anyways.

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c (hunk 19 above)
Comment (petrie911) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:20]: I prefer to just have all prototypes at the top, rather than determine which ones are needed.

Comment (petrie911) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:20]: Eh, scratch that

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c (hunk 21 above)
Comment (fig02) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:124]: yeah i prefer parens myself. its a topic a bunch of us disagree on lmao
but yeah petrie is right, these will be defines anyway so in this specific case i wouldnt worry

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h [hunk 24]
```diff
@@ -10,10 +10,29 @@ typedef void (*BgGndDarkmeiroActionFunc)(struct BgGndDarkmeiro*, GlobalContext*)
 
 typedef struct BgGndDarkmeiro {
     /* 0x0000 */ DynaPolyActor dyna;
-    /* 0x0164 */ char unk_164[0x8];
+    /* 0x0164 */ u16 actionState;    /* Bitflags for whether a block or timer is on.
+                                        0x8 for timer 2, 0x4 for timer 1, 0x2 for blocks.   */
```
Comment (fig02) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h:14]: code looks good to me.
just not sure on the name of this var. at the very least it should have the word flags in it imo

and action state sounds redundant. do the bit flags actually correspond to the action func?
a common pattern for this is to call it behavior if its separate from the action funcs. But if its directly tied, then actionFlags is probably good

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h [hunk 25]
```diff
@@ -10,10 +10,29 @@ typedef void (*BgGndDarkmeiroActionFunc)(struct BgGndDarkmeiro*, GlobalContext*)
...
+    /* 0x0166 */ s16 timer1;         /* Timer for switch flag N+1. Blocks use it to count 
+                                        down to vanishing when their switch flag is cleared.*/
+    /* 0x0168 */ s16 timer2;         /* Timer for switch flag N+2. Blocks use it to control
+                                        their transparency (does not work).                 */
```
Comment (fig02) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h:18]: in general i think itd be beneficial to keep these comments short and to the point. i personally aim to try to fit them on one line with a // comment (same for the enum below if possible)
not a hard guideline to go by, just a personal preference of mine and keeps documentation from getting too wordy

**Review by Dragorn421 — APPROVED** [review]

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h (hunk 24 above)
Comment (petrie911) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h:14]: I like actionFlags better than behavior. The flags keep track of what it's doing right now (counting down one or more timers or being on or off).

Comment (fig02) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.h:14]: do they correspond to the action function though? if they dont then they shouldnt have conflicting names

**Review by fig02 — APPROVED** [review]

**Review by ethteck — APPROVED** [review]

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 26]
```diff
@@ -1,7 +1,7 @@
 /*
  * File: z_bg_gnd_darkmeiro.c
  * Overlay: ovl_Bg_Gnd_Darkmeiro
- * Description: Clear block
+ * Description: Clear block and timer
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:4]:
```suggestion
 * Description: Shadow Trial actors (invisible path, clear block and timer)
```

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 27]
```diff
@@ -13,17 +13,15 @@
...
-void func_80878F34(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_80879128(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_80879134(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_808791BC(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_808791E8(BgGndDarkmeiro* this, GlobalContext* globalCtx);
-void func_80879334(BgGndDarkmeiro* this, GlobalContext* globalCtx);
+void BgGndDarkmeiro_ModeZero(BgGndDarkmeiro* this, GlobalContext* globalCtx);
+void BgGndDarkmeiro_BlockTimer(BgGndDarkmeiro* this, GlobalContext* globalCtx);
+void BgGndDarkmeiro_StaticBlock(BgGndDarkmeiro* this, GlobalContext* globalCtx);
+void BgGndDarkmeiro_SwitchBlock(BgGndDarkmeiro* this, GlobalContext* globalCtx);
```
[the remark this answers is not in the document]
Comment (petrie911) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:23]: Note that UpdateStaticBlock is _also_ a noop. Should I name them something like NoopPath and NoopBlock?

[the remark this answers is not in the document]
Comment (Roman971) [reply src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:23]: I think that one is meant to be dedicated to the "static block", whereas the other one is a default that never gets overridden in the switch in Init.

File: src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c [hunk 28]
```diff
@@ -35,25 +33,189 @@ const ActorInit Bg_Gnd_Darkmeiro_InitVars = {
...
+            this->timer2 = (Math_Coss((u16)this->timer1 * 0x1000) * 64.0f) + 127.0f;
+            if (this->timer2 > 127) {
+                this->timer2 = 127;
+            }
+        } else {
+            this->timer2 = vanishTimer * 8;
+        }
+
+        OPEN_DISPS(globalCtx->state.gfxCtx, "../z_bg_gnd_darkmeiro.c", 378);
+        //@ bug: Due to a bug in the display list, the transparency data is not used.
```
Comment (Roman971) [inline src/overlays/actors/ovl_Bg_Gnd_Darkmeiro/z_bg_gnd_darkmeiro.c:207]: The syntax needs to be fixed here if we want the bug annotation to be detected properly
```suggestion
        // @bug Due to a bug in the display list, the transparency data is not used.
```

**Review by Roman971 — APPROVED** [review]
